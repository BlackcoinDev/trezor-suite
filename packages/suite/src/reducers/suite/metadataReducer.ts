import produce from 'immer';
import { STORAGE, METADATA } from '@suite-actions/constants';
import { Action } from '@suite-types';
import { MetadataState } from '@suite-types/metadata';

export const initialState: MetadataState = {
    // is Suite trying to load metadata (get master key -> sync cloud)?
    enabled: false,
    initiating: false,
};

const metadataReducer = (state = initialState, action: Action): MetadataState =>
    produce(state, draft => {
        switch (action.type) {
            case STORAGE.LOAD:
                return action.payload.metadata || state;
            case METADATA.ENABLE:
                draft.enabled = true;
                break;
            case METADATA.DISABLE:
                draft.enabled = false;
                break;
            case METADATA.SET_PROVIDER:
                draft.provider = action.payload;
                break;
            case METADATA.SET_EDITING:
                draft.editing = action.payload;
                break;
            case METADATA.SET_INITIATING:
                draft.initiating = action.payload;
                break;
            case METADATA.SET_DATA:
                // todo: not here probably, does't matter, so far only PoC
                if (!draft.data) {
                    draft.data = {};
                }
                if (!draft.data[action.payload.provider]) {
                    draft.data[action.payload.provider] = {};
                }
                draft.data[action.payload.provider] = {
                    ...action.payload.data,
                    ...draft.data[action.payload.provider],
                };
                break;
            // no default
        }
    });

export default metadataReducer;
