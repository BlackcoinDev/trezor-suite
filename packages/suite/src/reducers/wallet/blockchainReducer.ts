import { BlockchainNetworks } from '@suite-common/wallet-types';
import { prepareBlockchainReducer } from '@suite-common/wallet-blockchain';
import { extraDependencies } from '../../support/extraDependecies';

export type BlockchainState = BlockchainNetworks;

const blockchainReducer = prepareBlockchainReducer(extraDependencies);
export default blockchainReducer;
