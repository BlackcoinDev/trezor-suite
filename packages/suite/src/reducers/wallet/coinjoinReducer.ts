import produce from 'immer';
import { ActiveRound } from '@trezor/coinjoin';
import { STORAGE } from '@suite-actions/constants';
import * as COINJOIN from '@wallet-actions/constants/coinjoinConstants';
import { Account, CoinjoinAccount } from '@suite-common/wallet-types';
import { Action } from '@suite-types';
import { PartialRecord } from '@trezor/type-utils';

type Client = {
    status: any[]; // TODO: Rounds from @trezor/coinjoin
};

type CoinjoinState = {
    accounts: CoinjoinAccount[];
    clients: PartialRecord<Account['symbol'], Client>;
};

const initialState: CoinjoinState = {
    accounts: [],
    clients: {},
};

const createAccount = (
    draft: CoinjoinState,
    { account, targetAnonymity }: Extract<Action, { type: typeof COINJOIN.ACCOUNT_CREATE }>,
) => {
    const exists = draft.accounts.find(a => a.key === account.key);
    if (exists) return;
    draft.accounts.push({
        key: account.key,
        targetAnonymity,
        previousSessions: [],
    });
};

const updateTargetAnonymity = (
    draft: CoinjoinState,
    {
        key,
        targetAnonymity,
    }: Extract<Action, { type: typeof COINJOIN.ACCOUNT_UPDATE_TARGET_ANONYMITY }>,
) => {
    const account = draft.accounts.find(a => a.key === key);
    if (!account) return;
    account.targetAnonymity = targetAnonymity;
};

const createSession = (
    draft: CoinjoinState,
    action: Extract<Action, { type: typeof COINJOIN.ACCOUNT_AUTHORIZE_SUCCESS }>,
) => {
    const account = draft.accounts.find(a => a.key === action.account.key);
    if (!account) return;
    account.session = {
        ...action.params,
        timeCreated: Date.now(),
        // phase: 0,
        deadline: Date.now(),
        registeredUtxos: [],
        signedRounds: [],
    };
};

const updateSession = (draft: CoinjoinState, accountKey: string, round: ActiveRound) => {
    const account = draft.accounts.find(a => a.key === accountKey);
    if (!account || !account.session) return;

    const deadline =
        round.phase !== 0
            ? new Date(Date.now() + 1000 * 60).toString()
            : round.roundParameters.inputRegistrationEnd;

    account.session = {
        ...account.session,
        phase: round.phase,
        deadline,
    };
};

const updateRegisteredUtxos = (
    draft: CoinjoinState,
    accountKey: string,
    registeredUtxos: string[],
) => {
    const account = draft.accounts.find(a => a.key === accountKey);
    if (!account || !account.session) return;

    account.session = {
        ...account.session,
        registeredUtxos,
    };
};

const signSession = (draft: CoinjoinState, accountKey: string, roundId: string) => {
    const account = draft.accounts.find(a => a.key === accountKey);
    if (!account || !account.session) return;
    account.session = {
        ...account.session,
        signedRounds: account.session.signedRounds.concat(roundId),
    };
};

const completeSession = (
    draft: CoinjoinState,
    action: Extract<Action, { type: typeof COINJOIN.SESSION_COMPLETED }>,
) => {
    const account = draft.accounts.find(a => a.key === action.accountKey);
    if (!account) return;
    if (account.session) {
        account.previousSessions.push({
            ...account.session,
            timeEnded: Date.now(),
        });
        delete account.session;
    }
};

const stopSession = (
    draft: CoinjoinState,
    action: Extract<Action, { type: typeof COINJOIN.ACCOUNT_UNREGISTER }>,
) => {
    const account = draft.accounts.find(a => a.key === action.account.key);
    if (!account) return;
    if (account.session) {
        account.previousSessions.push({
            ...account.session,
            timeEnded: Date.now(),
        });
        delete account.session;
    }
};

const saveCheckpoint = (
    draft: CoinjoinState,
    action: Extract<Action, { type: typeof COINJOIN.ACCOUNT_DISCOVERY_PROGRESS }>,
) => {
    const account = draft.accounts.find(a => a.key === action.account.key);
    if (!account) return;
    account.checkpoint = action.progress.checkpoint;
};

const createClient = (
    draft: CoinjoinState,
    action: Extract<Action, { type: typeof COINJOIN.CLIENT_ENABLE_SUCCESS }>,
) => {
    const exists = draft.clients[action.symbol];
    if (exists) return;
    draft.clients[action.symbol] = {
        status: [],
    };
};

export const coinjoinReducer = (
    state: CoinjoinState = initialState,
    action: Action,
): CoinjoinState =>
    produce(state, draft => {
        switch (action.type) {
            case STORAGE.LOAD:
                // Temporary code
                // coinjoin reducer is not stored in DB yet
                // restore accounts with coinjoin accountType
                action.payload.accounts.forEach(account => {
                    if (account.accountType === 'coinjoin') {
                        draft.accounts.push({
                            key: account.key,
                            targetAnonymity: 0,
                            previousSessions: [],
                        });
                    }
                });
                break;
            case COINJOIN.ACCOUNT_CREATE:
                createAccount(draft, action);
                break;
            case COINJOIN.ACCOUNT_UPDATE_TARGET_ANONYMITY:
                updateTargetAnonymity(draft, action);
                break;
            case COINJOIN.ACCOUNT_UPDATE_REGISTERED_UTXOS:
                updateRegisteredUtxos(draft, action.key, action.utxos);
                break;
            case COINJOIN.ACCOUNT_AUTHORIZE_SUCCESS:
                createSession(draft, action);
                break;
            case COINJOIN.ACCOUNT_UNREGISTER:
                stopSession(draft, action);
                break;
            case COINJOIN.ACCOUNT_DISCOVERY_PROGRESS:
                saveCheckpoint(draft, action);
                break;

            case COINJOIN.CLIENT_ENABLE_SUCCESS:
                createClient(draft, action);
                break;

            case COINJOIN.ROUND_PHASE_CHANGED:
                updateSession(draft, action.accountKey, action.round);
                break;

            case COINJOIN.ROUND_TX_SIGNED:
                signSession(draft, action.accountKey, action.roundId);
                break;

            case COINJOIN.SESSION_COMPLETED:
                completeSession(draft, action);
                break;

            // no default
        }
    });
