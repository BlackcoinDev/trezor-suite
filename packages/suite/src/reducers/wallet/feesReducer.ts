import produce from 'immer';
import { BLOCKCHAIN } from '@wallet-actions/constants';
import { NETWORKS } from '@wallet-config';
import { WalletAction } from '@wallet-types';
import { NetworksFeeInfo } from '@suite-common/wallet-types';

// type Symbol = Network['symbol'] | 'erc20';
export type State = NetworksFeeInfo;

const initialStatePredefined: Partial<State> = {
    // erc20: {
    //     blockHeight: 0,
    //     blockTime: 10,
    //     minFee: 1,
    //     maxFee: 100,
    //     levels: [{ label: 'normal', feePerUnit: '1', blocks: 0 }],
    // },
};

// fill initial state, those values will be changed by BLOCKCHAIN.UPDATE_FEE action
export const initialState = NETWORKS.reduce((state, network) => {
    if (network.accountType) return state;
    state[network.symbol] = {
        blockHeight: 0,
        blockTime: 10,
        minFee: 1,
        maxFee: 100,
        levels: [{ label: 'normal', feePerUnit: '1', blocks: 0 }],
    };
    return state;
}, initialStatePredefined as State);

const feesReducer = (state: State = initialState, action: WalletAction) =>
    produce(state, draft => {
        switch (action.type) {
            case BLOCKCHAIN.UPDATE_FEE:
                return {
                    ...draft,
                    ...action.payload,
                };
            // no default
        }
    });

export default feesReducer;
