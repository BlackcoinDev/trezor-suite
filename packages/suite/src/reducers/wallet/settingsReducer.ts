import produce from 'immer';
import { PROTO } from '@trezor/connect';
import { WalletSettings } from '@suite-common/wallet-types';
import { Action, AppState } from '@suite-types';

export type State = WalletSettings;

export const initialState: State = {
    localCurrency: 'usd',
    discreetMode: false,
    enabledNetworks: ['btc'],
    bitcoinAmountUnit: PROTO.AmountUnit.BITCOIN,
    lastUsedFeeLevel: {},
};

const settingsReducer = (state: State = initialState, action: Action): State =>
    produce(state, draft => {
        switch (
            action.type
            // no default
        ) {
        }
    });

export const selectEnabledNetworks = (state: AppState) => state.wallet.settings.enabledNetworks;

export default settingsReducer;
