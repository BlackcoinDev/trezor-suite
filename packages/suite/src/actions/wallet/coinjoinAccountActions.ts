import TrezorConnect from '@trezor/connect';
import * as COINJOIN from './constants/coinjoinConstants';
import { goto } from '../suite/routerActions';
import { addToast } from '../suite/notificationActions';
import { initCoinjoinClient, getCoinjoinClient } from './coinjoinClientActions';
import { CoinjoinBackendService } from '@suite/services/coinjoin/coinjoinBackend';
import { Dispatch, GetState } from '@suite-types';
import { Network } from '@suite-common/wallet-config';
import { Account, CoinjoinSessionParameters } from '@suite-common/wallet-types';
import { accountsActions, transactionsActions } from '@suite-common/wallet-core';
import {
    isAccountOutdated,
    analyzeTransactions,
    getAccountTransactions,
    getOriginalTransaction,
} from '@suite-common/wallet-utils';

const coinjoinAccountCreate = (account: Account) =>
    ({
        type: COINJOIN.ACCOUNT_CREATE,
        account,
    } as const);

const coinjoinAccountAuthorize = (account: Account) =>
    ({
        type: COINJOIN.ACCOUNT_AUTHORIZE,
        account,
    } as const);

const coinjoinAccountAuthorizeSuccess = (account: Account, params: CoinjoinSessionParameters) =>
    ({
        type: COINJOIN.ACCOUNT_AUTHORIZE_SUCCESS,
        account,
        params,
    } as const);

const coinjoinAccountAuthorizeFailed = (account: Account, error: string) =>
    ({
        type: COINJOIN.ACCOUNT_AUTHORIZE_FAILED,
        account,
        error,
    } as const);

const coinjoinAccountUnregister = (account: Account) =>
    ({
        type: COINJOIN.ACCOUNT_UNREGISTER,
        account,
    } as const);

export type CoinjoinAccountAction =
    | ReturnType<typeof coinjoinAccountCreate>
    | ReturnType<typeof coinjoinAccountAuthorize>
    | ReturnType<typeof coinjoinAccountAuthorizeSuccess>
    | ReturnType<typeof coinjoinAccountAuthorizeFailed>
    | ReturnType<typeof coinjoinAccountUnregister>;

const getKnownState = (
    account: Extract<Account, { backendType: 'coinjoin' }>,
    getState: GetState,
) =>
    account.discoveryCheckpoint && {
        blockHash: account.discoveryCheckpoint.blockHash,
        receiveCount: account.addresses
            ? account.addresses.unused.length + account.addresses.used.length
            : 0,
        changeCount: account.addresses?.change.length ?? 0,
        transactions: getAccountTransactions(
            account.key,
            getState().wallet.transactions.transactions,
        ).map(getOriginalTransaction),
    };

export const fetchAndUpdateAccount =
    (account: Account) => async (dispatch: Dispatch, getState: GetState) => {
        if (account.backendType !== 'coinjoin' || account.discoveryStatus === 'syncing') return;

        const knownState = getKnownState(account, getState);

        try {
            const api = CoinjoinBackendService.getInstance(account.symbol);
            if (!api) throw new Error('CoinjoinBackendService api not found');

            const { checkpoint, ...accountInfo } = await api.getAccountInfo({
                descriptor: account.descriptor,
                knownState,
            });

            // TODO add isPending check?
            if (!isAccountOutdated(account, accountInfo)) {
                if (account.discoveryStatus === 'initial') {
                    dispatch(
                        accountsActions.updateAccount({ ...account, discoveryStatus: 'ready' }),
                    );
                }
                return;
            }

            const analyze = analyzeTransactions(
                accountInfo.history.transactions || [],
                knownState?.transactions || [],
            );

            if (analyze.remove.length > 0) {
                dispatch(transactionsActions.removeTransaction({ txs: analyze.remove, account }));
            }
            if (analyze.add.length > 0) {
                dispatch(
                    transactionsActions.addTransaction({
                        transactions: analyze.add.reverse(),
                        account,
                    }),
                );
            }

            // TODO notify about analyze.newTransactions

            // get fresh info from reducer
            const updatedAccount = getState().wallet.accounts.find(a => a.key === account.key);
            if (updatedAccount?.backendType === 'coinjoin') {
                // finalize
                dispatch(
                    accountsActions.updateAccount(
                        {
                            ...updatedAccount,
                            discoveryStatus: 'ready',
                            discoveryCheckpoint: checkpoint,
                        },
                        accountInfo,
                    ),
                );
            }
        } catch (error) {
            // TODO
            console.warn('fetchAndUpdateAccount', error);
        }
    };

export const createCoinjoinAccount =
    (network: Network) => async (dispatch: Dispatch, getState: GetState) => {
        if (network.accountType !== 'coinjoin') {
            throw new Error('createCoinjoinAccount: invalid account type');
        }

        // initialize @trezor/coinjoin backend
        if (!CoinjoinBackendService.getInstance(network.symbol)) {
            await CoinjoinBackendService.createInstance(network.symbol);
        }

        // initialize @trezor/coinjoin client
        const client = await dispatch(initCoinjoinClient(network.symbol));
        if (!client) {
            return;
        }

        const { device } = getState().suite;

        // TODO: Disable safety_checks until Trezor FW will implement slip-0025
        // const safety =
        //     device?.features?.safety_checks !== 'PromptTemporarily'
        //         ? 'PromptTemporarily'
        //         : undefined;
        const experimentalFeatures = await TrezorConnect.applySettings({
            device,
            experimental_features: true,
            // safety_checks: safety,
        });
        if (!experimentalFeatures.success) {
            dispatch(addToast({ type: 'error', error: 'Experimental features not enabled' }));
            return;
        }

        // temporary hardcoded until Trezor FW will implement slip-0025
        const PATH = `m/86'/1'/25'`; // `m/10025'/1'/0'`  network.bip43Path

        // get coinjoin account xpub
        const publicKey = await TrezorConnect.getPublicKey({
            device,
            useEmptyPassphrase: device?.useEmptyPassphrase,
            path: PATH,
            coin: network.symbol,
        });
        if (!publicKey.success) {
            dispatch(addToast({ type: 'error', error: 'Public key not given' }));
            return;
        }

        // create empty account
        const account = dispatch(
            accountsActions.createAccount(
                device!.state!,
                {
                    index: 0,
                    path: PATH,
                    accountType: network.accountType,
                    networkType: network.networkType,
                    backendType: 'coinjoin',
                    coin: network.symbol,
                    derivationType: 0,
                    discoveryStatus: 'initial',
                },
                {
                    addresses: { change: [], used: [], unused: [] },
                    availableBalance: '0',
                    balance: '0',
                    descriptor: publicKey.payload.xpubSegwit || publicKey.payload.xpub,
                    empty: true,
                    history: { total: 0, unconfirmed: 0 },
                    legacyXpub: publicKey.payload.xpub,
                    page: { index: 1, size: 25, total: 1 },
                    utxo: [],
                },
            ),
        );
        dispatch(coinjoinAccountCreate(account.payload));

        // switch to account
        dispatch(
            goto('wallet-index', {
                params: {
                    symbol: network.symbol,
                    accountType: network.accountType,
                    accountIndex: 0,
                },
            }),
        );

        // start discovery
        dispatch(fetchAndUpdateAccount(account.payload));
    };

const authorizeCoinjoin =
    (account: Account, params: CoinjoinSessionParameters & { coordinator: string }) =>
    async (dispatch: Dispatch, getState: GetState) => {
        // initialize @trezor/coinjoin client
        const client = await dispatch(initCoinjoinClient(account.symbol));
        if (!client) {
            return;
        }

        const { device } = getState().suite;

        // authorize coinjoin session on Trezor
        dispatch(coinjoinAccountAuthorize(account));

        const auth = await TrezorConnect.authorizeCoinJoin({
            device,
            useEmptyPassphrase: device?.useEmptyPassphrase,
            path: account.path,
            coin: account.symbol,
            ...params,
        });

        if (auth.success) {
            dispatch(coinjoinAccountAuthorizeSuccess(account, params));
            return true;
        }

        dispatch(coinjoinAccountAuthorizeFailed(account, auth.payload.error));

        dispatch(
            addToast({
                type: 'error',
                error: `Coinjoin not authorized: ${auth.payload.error}`,
            }),
        );
    };

// called from coinjoin account UI
export const startCoinjoinSession =
    (account: Account, params: CoinjoinSessionParameters) => async (dispatch: Dispatch) => {
        if (account.accountType !== 'coinjoin') {
            throw new Error('startCoinjoinSession: invalid account type');
        }

        // initialize @trezor/coinjoin client
        const client = await dispatch(initCoinjoinClient(account.symbol));
        if (!client) {
            return;
        }

        // authorize CoinjoinSession on Trezor
        const authResult = await dispatch(
            authorizeCoinjoin(account, {
                ...params,
                coordinator: client.settings.coordinatorName,
            }),
        );

        // register authorized account
        if (authResult) {
            client.registerAccount(account);
        }
    };

// called from coinjoin account UI or exceptions like device disconnection, forget wallet/account etc.
export const stopCoinjoinSession = (account: Account) => (dispatch: Dispatch) => {
    // get @trezor/coinjoin client if available
    const client = dispatch(getCoinjoinClient(account.symbol));
    if (!client) {
        return;
    }

    // unregister account in @trezor/coinjoin
    client.unregisterAccount(account.key);

    // dispatch data to reducer
    dispatch(coinjoinAccountUnregister(account));
};
