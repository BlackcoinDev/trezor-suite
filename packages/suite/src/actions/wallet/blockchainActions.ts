import {
    initThunk,
    onBlockMinedThunk,
    onConnectThunk,
    onDisconnectThunk,
    onNotificationThunk,
    preloadFeeInfoThunk,
    reconnectThunk,
    setCustomBackendThunk,
    subscribeThunk,
    syncAccountsThunk,
    unsubscribeThunk,
    updateFeeInfoThunk,
} from '@suite-common/wallet-blockchain';
import { State as FeeState } from '@wallet-reducers/feesReducer';

import { BLOCKCHAIN } from './constants';

import type { Network, CustomBackend, BackendType } from '@wallet-types';
import type { Timeout } from '@trezor/type-utils';
// Conditionally subscribe to blockchain backend
// called after TrezorConnect.init successfully emits TRANSPORT.START event
// checks if there are discovery processes loaded from LocalStorage
// if so starts subscription to proper networks

export type BlockchainAction =
    | {
          type: typeof BLOCKCHAIN.CONNECTED;
          payload: Network['symbol'];
      }
    | {
          type: typeof BLOCKCHAIN.RECONNECT_TIMEOUT_START;
          payload: {
              symbol: Network['symbol'];
              id: Timeout;
              time: number;
              count: number;
          };
      }
    | {
          type: typeof BLOCKCHAIN.UPDATE_FEE;
          payload: Partial<FeeState>;
      }
    | {
          type: typeof BLOCKCHAIN.SYNCED;
          payload: {
              symbol: Network['symbol'];
              timeout: Timeout;
          };
      }
    | {
          type: typeof BLOCKCHAIN.SET_BACKEND;
          payload:
              | CustomBackend
              | {
                    coin: Network['symbol'];
                    type: 'default';
                };
      };

export const preloadFeeInfo = preloadFeeInfoThunk;

export const updateFeeInfo = updateFeeInfoThunk;

export const reconnect = reconnectThunk;

export const setCustomBackend = setCustomBackendThunk;

export const resetBackend = (coin: Network['symbol']): BlockchainAction => ({
    type: BLOCKCHAIN.SET_BACKEND,
    payload: {
        coin,
        type: 'default',
    },
});

export const setBackend = (
    coin: Network['symbol'],
    type: BackendType,
    urls: string[],
): BlockchainAction => ({
    type: BLOCKCHAIN.SET_BACKEND,
    payload: {
        coin,
        type,
        urls,
    },
});

export const init = initThunk;

// called from WalletMiddleware after ACCOUNT.ADD/UPDATE action
// or after BLOCKCHAIN.CONNECT event (blockchainActions.onConnect)
export const subscribe = (symbol: Network['symbol'], fiatRates = false) =>
    subscribeThunk({ symbol, fiatRates });

// called from WalletMiddleware after ACCOUNT.REMOVE action
export const unsubscribe = unsubscribeThunk;

export const syncAccounts = syncAccountsThunk;

export const onConnect = onConnectThunk;

export const onBlockMined = onBlockMinedThunk;

export const onNotification = onNotificationThunk;

export const onDisconnect = onDisconnectThunk;
