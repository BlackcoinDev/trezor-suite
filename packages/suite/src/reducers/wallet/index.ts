import { combineReducers } from 'redux';
import graphReducer from './graphReducer';
import discoveryReducer from './discoveryReducer';
import {
    prepareAccountsReducer,
    prepareFiatRatesReducer,
    prepareTransactionsReducer,
    prepareBlockchainReducer,
    prepareWalletSettingsReducer,
} from '@suite-common/wallet-core';
import selectedAccountReducer from './selectedAccountReducer';
import receiveReducer from './receiveReducer';
import feesReducer from './feesReducer';
import coinmarketReducer from './coinmarketReducer';
import sendFormReducer from './sendFormReducer';
import accountSearchReducer from './accountSearchReducer';
import formDraftReducer from './formDraftReducer';
import cardanoStakingReducer from './cardanoStakingReducer';
import pollingReducer from './pollingReducer';
import { coinjoinReducer } from './coinjoinReducer';
import { extraDependencies } from '@suite/support/extraDependencies';

export const transactionsReducer = prepareTransactionsReducer(extraDependencies);
export const accountsReducer = prepareAccountsReducer(extraDependencies);
export const blockchainReducer = prepareBlockchainReducer(extraDependencies);
export const fiatRatesReducer = prepareFiatRatesReducer(extraDependencies);
export const settingsReducer = prepareWalletSettingsReducer(extraDependencies);

const WalletReducers = combineReducers({
    fiat: fiatRatesReducer,
    graph: graphReducer,
    settings: settingsReducer,
    transactions: transactionsReducer,
    discovery: discoveryReducer,
    accounts: accountsReducer,
    selectedAccount: selectedAccountReducer,
    receive: receiveReducer,
    fees: feesReducer,
    blockchain: blockchainReducer,
    coinmarket: coinmarketReducer,
    send: sendFormReducer,
    accountSearch: accountSearchReducer,
    formDrafts: formDraftReducer,
    cardanoStaking: cardanoStakingReducer,
    pollings: pollingReducer,
    coinjoin: coinjoinReducer,
});

export default WalletReducers;
