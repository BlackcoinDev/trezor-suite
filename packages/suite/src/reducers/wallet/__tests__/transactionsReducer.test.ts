import { STORAGE } from '../../../actions/suite/constants';
import { ACCOUNT, TRANSACTION } from '../../../actions/wallet/constants';
import reducer, { initialState } from '../transactionReducer';
import * as transactionActions from '../transactionActions';

const { getWalletTransaction } = global.JestMocks;

describe('transactions reducer', () => {
    // it('ACCOUNT.REMOVE action', () => {
    //     expect(
    //         reducer(undefined, {
    //             type: ACCOUNT.REMOVE,
    //             payload: [{ key: 'account1' }, { key: 'account2' }],
    //         }),
    //     ).toEqual(initialState);
    // });

    it('TRANSACTION.ADD action', () => {
        const account = global.JestMocks.getWalletAccount();

        expect(
            reducer(undefined, {
                type: TRANSACTION.ADD,
                transactions: [getWalletTransaction()],
                account,
            }),
        ).toMatchSnapshot();
    });

    it('TRANSACTION.REMOVE action', () => {
        const account1 = global.JestMocks.getWalletAccount({ descriptor: 'xpub1' });
        const account2 = global.JestMocks.getWalletAccount({ descriptor: 'xpub2' });
        const transaction1 = global.JestMocks.getWalletTransaction({
            blockTime: 1565792979,
            blockHeight: 5,
        });
        const transaction2 = global.JestMocks.getWalletTransaction({
            blockTime: 1565792379,
            blockHeight: 4,
        });

        const stateWithTransactions1 = reducer(undefined, {
            type: TRANSACTION.ADD,
            transactions: [transaction1],
            account: account1,
        });
        const stateWithTransactions2 = reducer(stateWithTransactions1, {
            type: TRANSACTION.ADD,
            transactions: [transaction1, transaction2],
            account: account2,
        });

        console.log(stateWithTransactions2.transactions[account2.key]);

        const stateWithRemovedTransaction = reducer(stateWithTransactions2, {
            type: TRANSACTION.REMOVE,
            txs: [transaction2],
            account: account2,
        });

        console.log(stateWithRemovedTransaction.transactions[account2.key]);

        expect(stateWithRemovedTransaction.transactions[account2.key].length).toBe(1);
        expect(stateWithRemovedTransaction.transactions[account1.key].length).toBe(1);
    });
});
