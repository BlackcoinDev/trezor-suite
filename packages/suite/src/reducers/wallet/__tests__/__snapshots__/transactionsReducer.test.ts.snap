// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transactions reducer TRANSACTION.ADD action 1`] = `
Object {
  "error": null,
  "isLoading": false,
  "transactions": Object {
    "0xFA01a39f8Abaeb660c3137f14A310d0b414b2A15-eth-7dcccffe70d8bb8bb28a2185daac8e05639490eee913b326097ae1d73abc8b4f": Array [
      Object {
        "amount": "0.00001",
        "blockHash": "00000000000000000017277948d61a631dae6cce1d7fb501301b825599189f51",
        "blockHeight": 590093,
        "blockTime": 1565797979,
        "descriptor": "zpub6rszzdAK6RuafeRwyN8z1cgWcXCuKbLmjjfnrW4fWKtcoXQ8787214pNJjnBG5UATyghuNzjn6Lfp5k5xymrLFJnCy46bMYJPyZsbpFGagT",
        "details": Object {
          "size": 225,
          "totalInput": "0.80720012",
          "totalOutput": "0.80719868",
          "vin": Array [
            Object {
              "addresses": Array [
                "tb1q4nytpy37cuz8yndtfqpau4nzsva0jh787ny3yg",
              ],
              "isAddress": true,
              "n": 0,
              "sequence": 4294967294,
              "txid": "c894b064beb2f9be4b0d64cffcd89da2e8dc6decac399f5617323a303e07e4e1",
              "value": "0.80720012",
            },
          ],
          "vout": Array [
            Object {
              "addresses": Array [
                "tb1q4s560ew83wcd6lcjg7uku9qlx4p6gwh74q4jap",
              ],
              "hex": "0014ac29a7e5c78bb0dd7f1247b96e141f3543a43afe",
              "isAddress": true,
              "n": 0,
              "value": "0.80718868",
            },
            Object {
              "addresses": Array [
                "mvbu1Gdy8SUjTenqerxUaZyYjmveZvt33q",
              ],
              "hex": "76a914a579388225827d9f2fe9014add644487808c695d88ac",
              "isAddress": true,
              "n": 1,
              "value": "0.00001",
            },
          ],
        },
        "deviceState": "7dcccffe70d8bb8bb28a2185daac8e05639490eee913b326097ae1d73abc8b4f",
        "fee": "0.00000144",
        "symbol": "btc",
        "targets": Array [
          Object {
            "addresses": Array [
              "mvbu1Gdy8SUjTenqerxUaZyYjmveZvt33q",
            ],
            "amount": "0.00001",
            "isAccountTarget": true,
            "isAddress": true,
            "n": 1,
          },
        ],
        "tokens": Array [],
        "totalSpent": "0.00001144",
        "txid": "7e58757f43015242c0efa29447bea4583336f2358fdff587b52bbe040ad8982a",
        "type": "sent",
      },
    ],
  },
}
`;
